[WFLY-21572] Documentation for HashiCorp Vault as credential source (#19721)
wildfly/wildfly: docs/src/main/asciidoc/_elytron/HashiCorp_Vault_Integration.adoc

There are several ways to install the HashiCorp Vault feature pack. Refer to the Galleon documentation for these.

== HashiCorp Vault credential store

After the feature pack is installed, define one or more credential store resources in the `hashicorp-vault` subsystem. You can then reference credentials from these stores elsewhere in the configuration using the standard `credential-reference` pattern with `store` and `alias`.

Aliases use the form _`<vault-path>.<key>`_: the path to the secret in Vault (e.g. `secret/database`) followed by a dot and the key name (e.g. `password`), giving `secret/database.password`.

=== Credential references backed by HasihCorp Vault

Use the credential store name and alias in any resource that supports `credential-reference` (e.g. datasources, Elytron key stores, etc.):

[source,xml,options="nowrap"]
----

> <credential-reference store="my-vault" alias="secret/database.password"/>

Reviewer: This LGTM, since it's a demonstration of the resulting config, with the mechanism to achieve that not being the focus.

A future improvement would be a concrete example of setting such a config (e.g. on a datasource). It's covered elsewhere, but some redundancy is not terrible.

[[Vault_Integration]]
= HashiCorp Vault Integration

> To use credential stores backed by HashiCorp Vault in the WildFly management model, the Galleon Vault feature pack can be used. This feature pack adds the `hashicorp-vault` subsystem to WildFly, which allows you to define Elytron credential store resources that read secrets from a HashiCorp Vault server. It also allows you to define expressions backed by HashiCorp Vault.

Reviewer: ```suggestion
To use credential stores backed by HashiCorp Vault in the WildFly management model, the Vault feature-pack can be used. This feature-pack adds the `hashicorp-vault` subsystem to WildFly, which allows you to define Elytron credential store resources that read secrets from a HashiCorp Vault server. It also allows you to define expressions backed by HashiCorp Vault.
```

Author: resolved

> There are several ways to install the Vault feature pack. Refer to the Galleon documentation for these.

Reviewer: ```suggestion
There are several ways to install the Vault feature-pack. Refer to the xref:WildFly_Maven_Plugin_Guilde{outfilesuffix}[WildFly Maven Plugin Guide] and the xref:Galleon_Guide{outfileSuffix}[Galleon Provisioning Guide] for details.
```

Author: resolved

== HashiCorp Vault credential store

> After the feature pack is installed, define one or more credential store resources in the `hashicorp-vault` subsystem. You can then reference credentials from these stores elsewhere in the configuration using the standard `credential-reference` pattern with `store` and `alias`.

Reviewer: ```suggestion
After the feature-pack is installed, define one or more credential store resources in the `hashicorp-vault` subsystem. You can then reference credentials from these stores elsewhere in the configuration using the standard `credential-reference` pattern with `store` and `alias`.
```

Author: resolved
